### ISSUE TITLE
"android" recipe isn't compatible with Python 3

### ISSUE THREAD

**Issue by germn** [kivy/python-for-android#1093]
apk built with "android" recipe and Python 3 will crash after starup:

> 06-05 12:45:47.101 30382 30533 I python  :    File "/data/user/0/org.voctrain.voctrain/files/app/crystax_python/site-packages/kivy/base.py", line 210, in remove_android_splash
> 06-05 12:45:47.111 30382 30533 I python  :      from android import remove_presplash
> 06-05 12:45:47.111 30382 30533 I python  :    File "/data/user/0/org.voctrain.voctrain/files/app/crystax_python/site-packages/android/__init__.py", line 8, in <module>
> 06-05 12:45:47.111 30382 30533 I python  :      from android._android import *
> 06-05 12:45:47.111 30382 30533 I python  :    File "android/_android.pyx", line 178, in init android._android (android/_android.c:6886)
> 06-05 12:45:47.111 30382 30533 I python  :  TypeError: can't concat bytes to str
> 06-05 12:45:47.391 30382 30533 I python  : Python for android ended.

Originally reported [here](https://groups.google.com/forum/#!topic/kivy-users/CgGTCN1TG2w), I faced same issue.

**Comment by inclement** [comment]
Thanks for the report, I'll look into it, I guess it shouldn't be hard to fix.

For what it's worth, the android module isn't essential, and most of its functionality can be accomplished easily using pyjnius, so you may be able to use that as a workaround. That is probably why I haven't noticed this before.

**Comment by inclement** [comment]
I looked into this but couldn't reproduce it. Am I right in thinking that the crash should occur immediately on import of the android module?

**Comment by germn** [comment]
@inclement I didn't import `android` module at all. I add `android` recipe as dependency only [to avoid](https://github.com/kivy/python-for-android/issues/963) logging message "Could not remove android presplash".

I also use this code inside app's build method, it's probably related:

```
if platform == "android":
    from jnius import autoclass
    activity = autoclass('org.kivy.android.PythonActivity').mActivity
    activity.removeLoadingScreen()
```

Sorry for not mentioning it's earlier.

**Comment by inclement** [comment]
I see, there shouldn't be anything special there (in fact, I remove the presplash this way myself). There are no disadvantages to just not including the android module and letting Kivy give the warning.

Looking at the code, I don't actually see why this error should occur, as neither of the strings being combined looks like bytes.

**Comment by germn** [comment]
From traceback error happens [here](https://github.com/kivy/python-for-android/blob/master/pythonforandroid/recipes/android/src/android/_android.pyx#L178):

> python_act = autoclass(JAVA_NAMESPACE + '.PythonActivity')

Probably `JAVA_NAMESPACE` is `bytes` on both Python 2 and 3, while `'.PythonActivity'` is `bytes` on Python 2 and `str` on Python 3.

I'm sorry, I can't fix/test it right now, if you want I can check and if it's right add PR later, after few days.

**Comment by inclement** [comment]
That's what I guessed, but JAVA_NAMESPACE is set in config.pxi, and is just a normal string. Maybe it's implicitly bytes, but that's not what I had expected. I did also test this, expecting to fix the error that way, but that's when I couldn't reproduce the problem.

If you can reproduce it and a fix of this nature seems to work, a PR would certainly be welcome though :)

**Comment by bbregeault** [comment]
I hit that bug too and it appeared to come from the cython version I use (0.23.4).

Cython 0.23.4 treats strings differently whether they are in a DEF statement or not.
So `JAVA_NAMESPACE` is `bytes` because it is set in a DEF statement in config.pxi and `'.PythonActivity'` is `str`.

Evidence can be found in the C file `android/_android.c` that cython generates from `android/_android.pyx`.
In my case, here are the relevant lines for the string declarations : 

```c
static __Pyx_StringTabEntry __pyx_string_tab[] = {
  ...
  {&__pyx_kp_s_PythonActivity, __pyx_k_PythonActivity, sizeof(__pyx_k_PythonActivity), 0, 0, 1, 0},
  ...
  {&__pyx_kp_b_org_kivy_android, __pyx_k_org_kivy_android, sizeof(__pyx_k_org_kivy_android), 0, 0, 0, 0},
  ...
};
```
Note the prefixes : `__pyx_kp_b` for `bytes` and `__pyx_kp_s` for `str`.
Accordingly, the parameter that differs (1 or 0) is called `is_str`in the typedef declaration from the same file :
```c
typedef struct {PyObject **p; char *s; const Py_ssize_t n; const char* encoding;
                const char is_unicode; const char is_str; const char intern; } __Pyx_StringTabEntry;
```
So later, the concatenation obviously goes BOOM:

```c
__pyx_t_3 = PyNumber_Add(__pyx_kp_b_org_kivy_android, __pyx_kp_s_PythonActivity);
```

I found that cython version 0.28.5 fixes the problem but triggers other compilation errors in the kivy recipe.
With a lot of patience I found the following changes solved the bug with cython 0.23.4:
- force a unicode string in config.pxi : `DEF JAVA_NAMESPACE = u'org.kivy.android'`
- remove the `DEF` keyword in config.pxi
- hardcode the string as proposed [here](https://gist.github.com/kived/c833c97bbd044e68515e)
- force cython to generate level 3 code: `cython -3 _android.pyx`

Unfortunately I don't know the p4a internals well enough to be able to propose a PR on my own but I could use some guidance.
And maybe it's not worth it since it's a bug from an old cython version.
